ladybug_geometry/geometry2d/mesh.py

```diff
@@ -0,0 +1,500 @@
+        polygon can be subdivided into a set of triangles.
+
+        Args:
+            polygon: A Polygon2D.
+            purge: A boolean to indicate if duplicate vertices should be shared between
+                faces. This has no bearing on the triagnulation of convex polygons
+                and only affects concave polygons with more than 4 sides.
+                Default is True to purge duplicate vertices, which can be slow
+                for convace polygons composed of a large number of triangles but
+                ultimately results in a higher-quality mesh. Default is True.
+        """
+        assert isinstance(polygon, Polygon2D), 'polygon must be a Polygon2D to use ' \
+            'from_polygon_triangulated. Got {}.'.format(type(polygon))
+
+        if polygon.is_convex:
+            # super-fast fan triangulation!
+            _faces = []
+            for i in xrange(1, len(polygon) - 1):
+                _faces.append((0, i, i + 1))
+            _new_mesh = cls(polygon.vertices, _faces)
+        elif len(polygon) == 4:
+            _faces = Mesh2D._concave_quad_to_triangles(polygon.vertices)
+            _new_mesh = cls(polygon.vertices, _faces)
+        else:
+            # slow ear-clipping method
+            _faces = Mesh2D._ear_clipping_triangulation(polygon)
+            _new_mesh = cls.from_face_vertices(_faces, purge)
+
+        return _new_mesh
+
+    @classmethod
+    def from_polygon_grid(cls, polygon, x_dim, y_dim, generate_centroids=True):
+        """Initialize a gridded Mesh2D from a Polygon2D.
+
+        Note that this gridded mesh will likely not completely fill the polygon.
+        Essentially, this method generates a grid over the domain of the polygon
+        and then removes any points that do not lie within the polygon.
+
+        Args:
+            polygon: A Polygon2D object.
+            x_dim: The x dimension of the grid cells as a number.
+            y_dim: The y dimension of the grid cells as a number.
+            generate_centroids: Set to True to have the face centroids generated
+                alongside the grid of vertices, which is much faster than having
+                them generated upon request as they typically are. However, if you
+                have no need for the face centroids, you would save memory by setting
+                this to False. Default is True.
+        """
+        assert isinstance(polygon, Polygon2D), 'Expected Polygon2D for' \
+            ' Mesh2D.from_polygon_grid. Got {}'.format(type(polygon))
+        # figure out how many x and y cells to make
+        _x_dim, _num_x = Mesh2D._domain_dimensions(polygon.max.x - polygon.min.x, x_dim)
+        _y_dim, _num_y = Mesh2D._domain_dimensions(polygon.max.y - polygon.min.y, y_dim)
+
+        # for tolerance reasons, we shrink the x and y dimensions by a very small amount
+        # this avoids the fringe cases noted in the Polygon2d.is_point_inside description
```

Comment:
Can we do this reversed and expand the polygon? In that case we will ensure the the dimensions are kept the same as input and the issue with fringe cases will also go away.

Reply:
Let me think about it.  We might not understand the full implications of the suggestion.  For example, what point do you scale the polygon from?  Does it still avoid odd situations where you don't get the grid going all of the way to the edge when you expect it to?

Reply:
After thinking about it, I realized that we should get the same result by scaling the polygon from the center (not the centroid).  It will take a little longer to generate this scaled polygon but we can do it that way if you want to preserve the exact grid size for the few cases it might matter.  It's still going to be subject to python floating point tolerance and the input grid size usually does not perfectly match the generated grid size unless the dimensions of the bounding rectangle are perfectly divisible by the grid size.  All of this is to say that I am not sure that it will really be that different in practice but I will implement it as you say for the few cases where it may matter.

Reply:
Actually, it wasn't from the center that it needed to be scaled but rather it needed to be scaled from the min point and moved a very tiny distance.  Now it gives the same result with no change in grid size for tolerance.